books/RayTracingTheNextWeek.html

```diff
@@ -1825,21 +1858,21 @@
 Let’s make the 5 walls and the light of the box:
 
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++
-    hittable *cornell_box() {
-        hittable **list = new hitable*[5];
-        int i = 0;
-        material *red = new lambertian(new constant_texture(vec3(0.65, 0.05, 0.05)));
-        material *white = new lambertian(new constant_texture(vec3(0.73, 0.73, 0.73)));
-        material *green = new lambertian(new constant_texture(vec3(0.12, 0.45, 0.15)));
-        material *light = new diffuse_light(new constant_texture(vec3(15, 15, 15)));
+    hittable_list cornell_box() {
+        hittable_list objects;
 
-        list[i++] = new yz_rect(0, 555, 0, 555, 555, green);
-        list[i++] = new yz_rect(0, 555, 0, 555, 0, red);
-        list[i++] = new xz_rect(213, 343, 227, 332, 554, light);
-        list[i++] = new xz_rect(0, 555, 0, 555, 0, white);
-        list[i++] = new xy_rect(0, 555, 0, 555, 555, white);
+        auto red = new lambertian(new constant_texture(vec3(0.65, 0.05, 0.05)));
+        auto white = new lambertian(new constant_texture(vec3(0.73, 0.73, 0.73)));
+        auto green = new lambertian(new constant_texture(vec3(0.12, 0.45, 0.15)));
+        auto light = new diffuse_light(new constant_texture(vec3(15, 15, 15)));
 
-        return new hittable_list(list,i);
+        objects.add(new yz_rect(0, 555, 0, 555, 555, green));
+        objects.add(new yz_rect(0, 555, 0, 555, 0, red));
+        objects.add(new xz_rect(213, 343, 227, 332, 554, light));
+        objects.add(new xz_rect(0, 555, 0, 555, 0, white));
+        objects.add(new xy_rect(0, 555, 0, 555, 555, white));
```

Comment:
The sharing of materials between the objects in this scene is the #1 reason that rewriting the memory deletions is so hard
:((

Reply:
This change is largely about moving the dynamic list to using `std::vector`, and thus having defined destructor logic. It currently has bare `hittable` pointers. For the next rev, we can convert this to `std::vector<std::shared_ptr<hittable>>`, and then ensure that object/material creation returns a shared pointer.